lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2.cpp

```diff
@@ -584,55 +608,175 @@ Value loadArg(ConversionPatternRewriter &rewriter, Location loc, Value tensor,
-  Value warpN = urem(multiDimWarpId[1], i32_val(shapePerCTA[1] / 8));
-
+  Value warpB = urem(multiDimWarpId[0], i32_val(shapePerCTA[0]));
   int warpsPerTile;
+  auto rank = shapePerCTA.size();
+  Value warpM = urem(multiDimWarpId[1], i32_val(shapePerCTA[1] / 16));
+  Value warpN = urem(multiDimWarpId[2], i32_val(shapePerCTA[2] / 8));
   if (isA)
-    warpsPerTile = std::min<int>(warpsPerCTA[0], shapePerCTA[0] / 16);
-  else
     warpsPerTile = std::min<int>(warpsPerCTA[1], shapePerCTA[1] / 16);
-
-  std::function<void(int, int)> loadFn;
+  else
+    warpsPerTile = std::min<int>(warpsPerCTA[2], shapePerCTA[2] / 16);
+  std::function<void(int, int, int)> loadFn;
   if (isA)
     loadFn = getLoadMatrixFn(
-        tensor, smemObj, mmaLayout, warpsPerTile /*warpsPerTile*/, 1 /*kOrder*/,
-        kWidth, {mmaInstrM, mmaInstrK} /*instrShape*/,
-        {matShapeM, matShapeK} /*matShape*/, warpM /*warpId*/, lane /*laneId*/,
+        tensorTy, smemObj, mmaLayout, warpsPerTile /*warpsPerTile*/,
+        2 /*kOrder*/, kWidth, {1, mmaInstrM, mmaInstrK} /*instrShape*/,
+        {1, matShapeM, matShapeK} /*matShape*/,
+        {warpB, warpM, warpN} /*multiDimWarpId*/, lane /*laneId*/,
         vals /*vals*/, isA /*isA*/, typeConverter /* typeConverter */,
         rewriter /*rewriter*/, loc /*loc*/);
   else
     loadFn = getLoadMatrixFn(
-        tensor, smemObj, mmaLayout, warpsPerTile /*warpsPerTile*/, 0 /*kOrder*/,
-        kWidth, {mmaInstrK, mmaInstrN} /*instrShape*/,
-        {matShapeK, matShapeN} /*matShape*/, warpN /*warpId*/, lane /*laneId*/,
+        tensorTy, smemObj, mmaLayout, warpsPerTile /*warpsPerTile*/,
+        1 /*kOrder*/, kWidth, {1, mmaInstrK, mmaInstrN} /*instrShape*/,
+        {1, matShapeK, matShapeN} /*matShape*/,
+        {warpB, warpM, warpN} /*multiDimWarpId*/, lane /*laneId*/,
         vals /*vals*/, isA /*isA*/, typeConverter /* typeConverter */,
         rewriter /*rewriter*/, loc /*loc*/);
 
   // Perform loading.
-  int numRepOuter = isA ? numRep[0] : std::max<int>(numRep[1] / 2, 1);
-  int numRepK = isA ? numRep[1] : numRep[0];
-  for (int m = 0; m < numRepOuter; ++m)
-    for (int k = 0; k < numRepK; ++k)
-      loadFn(2 * m, 2 * k);
+  int numRepBatch = numRep[0];
+  int numRepOuter = isA ? numRep[1] : std::max<int>(numRep[2] / 2, 1);
+  int numRepK = isA ? numRep[2] : numRep[1];
+  for (int b = 0; b < numRepBatch; ++b)
+    for (int m = 0; m < numRepOuter; ++m)
+      for (int k = 0; k < numRepK; ++k)
+        loadFn(b, 2 * m, 2 * k);
 
   // Format the values to LLVM::Struct to passing to mma codegen.
-  return composeValuesToDotOperandLayoutStruct(vals, numRepOuter, numRepK,
-                                               typeConverter, loc, rewriter);
+  return composeValuesToDotOperandLayoutStruct(
+      vals, numRepBatch, numRepOuter, numRepK, typeConverter, loc, rewriter);
+}
+
+template <typename T>
+SmallVector<T> insertValue(ArrayRef<T> vec, unsigned index, T value) {
```

Comment:
Does `insertValue` have to be a template function?

Reply:
I'm using the same functions for inserting an `unsigned`/`int64_t`/`Value`

Reply:
Unless the confusion was about the name `insertValue` which might be understood as `mlir::Value` ? Maybe I should rename to `insertElement` ?

Reply:
It's ok, never mind